text/0000-semantic-private-in-public.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: semantic_private_in_public
+- Start Date: 2016-07-09
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Enforce that public APIs do not expose private definitions at the semantic level, while allowing the use of private aliases and blanket implementations for convenience and automation.
+
+# Motivation
+[motivation]: #motivation
+
+The "private-in-public" rules ensure the transitivity of abstraction. That is, one must be able to name types and bounds used in public APIs, from anywhere else.
```

Comment:
"able to name" is not entirely correct, "able to name after addition of arbitrary number of reexports" is closer to the truth. E.g.
```
mod m {
    mod detail {
        pub struct S;
    }
    pub fn f() -> detail::S { detail::S }
}
```
is allowed because `S` is *potentially* nameable outside of `m` (if reexported), but not *actually* nameable.
I'd like to avoid promising namebility in docs, including RFCs, because it's already one of the most common misconceptions about our privacy system.